### PR TITLE
 Implement Comprehensive Metrics Tracking System

### PR DESCRIPTION
**Pull request by Gyan172004**
## Overview

This pull request introduces a comprehensive metrics tracking system for the docFiller project, which includes the following changes:

- **Metrics Management**: Centralized management of metrics using a singleton pattern.
- **Analytics**: Calculation of key metrics such as API call performance and usage tracking.
- **UI Integration**: Added UI components to display and visualize metrics data in the options page.

### Metrics System Implementation
- **/src/utils/metricsCalculator.ts**: Core logic for calculating and processing metrics data.
- **/src/utils/metricsManager.ts**: Singleton for managing metrics data centrally across the application.
- **/src/background/index.ts**: Integrated metrics tracking for API calls and background processes.
- **/src/options/metrics.ts**: UI components for displaying and interacting with metrics data in the options page.

### Storage Integration
- **/src/utils/storage/metricsStorage.ts**: Handles persistent storage of metrics data.
- **/src/utils/storage/getMetrics.ts**: Functions for retrieving metrics data from storage.
- **/src/utils/storage/setMetrics.ts**: Functions for storing updated metrics data.

### LLM Engine Integration
- **/src/docFillerCore/engines/gptEngine.ts**: Added performance tracking hooks to measure LLM engine performance.
- **/src/utils/llmEngineTypes.ts**: New type definitions related to metrics and LLM engine.

### UI Components
- **/src/options/options.ts**: Added a section to the options page for displaying metrics data.
- **/src/components/MetricsCard.tsx**: A reusable component for visualizing metrics on the UI.
- **/src/styles/metrics.css**: Styling and layout for metrics UI components.

## Technical Notes

- **Rolling History**: All metrics data maintains a 14-day rolling history for performance and usage tracking.
- **Type Safety**: TypeScript interfaces are used to ensure type safety across the metrics system.
- **Real-time Updates**: Implemented real-time updates using the Chrome storage API and event listeners.

## Screenshot
Screenshot 2024-12-14 164428

---

Please review the changes and let me know if you have any questions or feedback.

### CODE REVIEW

**Comment by rootCircle** [comment]
This looks awesome! :star_struck:

File: src/utils/storage/metricsManager.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+// src/utils/storage/metricsManager.ts
+
+// Add missing interface
+import { LLMEngineType } from '@utils/llmEngineTypes';
+import { MetricsCalculator } from '@utils/metricsCalculator';
+
+export class MetricsManager {
+  private static instance: MetricsManager;
+  private static readonly STORAGE_KEY = 'metrics';
+  private static readonly MAX_HISTORY_DAYS = 14;
```
Comment (rootCircle) [inline src/utils/storage/metricsManager.ts:13]: how are we using this?

File: package-lock.json
Comment (rootCircle) [inline package-lock.json:1]: delete file, we use bun, also check if it's in gitignore or not. If not add package-lock.json to gitignore

File: src/docFillerCore/index.ts [hunk 2]
```diff
@@ -37,6 +41,24 @@ async function runDocFillerEngine() {
     }
   }
 
+  const llmConfig = await validateLLMConfiguration();
+  if (
```
Comment (rootCircle) [inline src/docFillerCore/index.ts]: can we make this if condition, a bit more readable?

use boolean variables instead?

Comment (rootCircle) [reply src/docFillerCore/index.ts]: I believe this must be about disallowing filling if config is broken?

Comment (rootCircle) [reply src/docFillerCore/index.ts]: add a helpful message in console as well

**Review by rootCircle — CHANGES_REQUESTED** [review]
I trust you :sweat_smile:

File: src/utils/storage/metricsManager.ts (hunk 1 above)
Comment (Gyan172004) [reply src/utils/storage/metricsManager.ts:13]: Its made to manage state for metrics and managing its properties like updating them

File: src/docFillerCore/index.ts (hunk 2 above)
Comment (Gyan172004) [reply src/docFillerCore/index.ts]: Figured out we already have correct invalid engine logic implemented in  `validateLLMConfiguration` handling both states of consensus engine .
Cleaned up the if condition, made it more readable, and added a helpful console message.

**Review by rootCircle — APPROVED** [review]

File: src/utils/storage/metricsManager.ts (hunk 1 above)
Comment (Gyan172004) [reply src/utils/storage/metricsManager.ts:13]: MAX_HISTORY_DAYS is to limit storage of data , everytime form is filled , we sort history data structure in recently first manner and then slice 14 days (2 week) history.

Comment (Gyan172004) [reply src/utils/storage/metricsManager.ts:13]: Better way can be we can reset it after total_forms>1000?

Comment (rootCircle) [reply src/utils/storage/metricsManager.ts:13]: We can create it to store metadata > 14 days via aggregation! Details of which are beyond scope of this PR!

**Pull request merged**